### DISCUSSION TITLE
EEG extraction

### DISCUSSION THREAD

**Discussion by PhD-GOAT** [raphaelvallat/yasa#84] [Q&A] [answered]
In this example (https://github.com/raphaelvallat/yasa/blob/master/notebooks/01_spindles_detection.ipynb), you can see the plotting using the txt file. But in Yasa, I wonder if there is a function to extract EEG every 30 seconds or if there is something to extract from mne.

data_N3_no-spindles_30sec_100Hz.txt

**Comment by umair-hassan** [comment, marked as answer]
MNE Raw data also works with YASA spindle_detect or sw_detect methods. For example see here:

yasa.spindles_detect(data, sf, ch_names=chan, multi_only=True, remove_outliers=True)

Here data is MNE raw date object
chan="C3" for example (depending on channel label in your imported raw data, similarly there are other flags).

I am posting an example code below from my notebooks: 

raws_c3 = raw.pick_channels(['C3'])
sp = yasa.spindles_detect(raws_c3, 200,ch_names="C3")
sp.summary()
sp.plot_average(ci=None, palette=['tab:grey']);
sp.summary()

**Reply by PhD-GOAT** [reply]
To be exact, I will read raw as an edf file, not a txt file. Looking at the command, it seems to pick one channel, but the division by zero error appears in yasa.spindles_detect.

**Reply by umair-hassan** [reply]
Yep I understand that you wont be using txt file but edf file instead, this should work for you:

import mne
import yasa
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from yasa import sleep_statistics


fname='XYZ.edf'
raw=mne.io.read_raw_edf(fname)
sampling_frequency=200;
sp = yasa.spindles_detect(raw, sampling_frequency,ch_names="LabelOfYourIntetrest")
sp.summary()
sp.plot_average(ci=None, palette=['tab:grey']);

**Reply by umair-hassan** [reply]
You will have to make sure the sampling frequency is as per your raw data in EDF and channel label is correct. 

We can instead just also get this automaticaly from info field of raw mne loaded data.

**Reply by PhD-GOAT** [reply]
Yeah, I did it on that command, too. print(raw.info['sfreq']) : 100 so i do 100Hz 

But division by zero error appears. From what I can tell, yasa doesn't work in all the edf files.

**Reply by umair-hassan** [reply]
Then possibly we can get the numpy array from the raw loaded data using raw.get_data() function of MNE

**Reply by PhD-GOAT** [reply]
```
import mne
import yasa
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from yasa import sleep_statistics

raw = mne.io.read_raw_edf('C:/Users/SC4001E0-PSG.edf', preload=True)
print(raw)
print(raw.ch_names)

data = raw.get_data()

raw_np = raw.pick_channels(['EEG Fpz-Cz'])
sp = yasa.spindles_detect(raw_np, 100, ch_names="EEG Fpz-Cz")
sp.summary()
sp.plot_average(ci=None, palette=['tab:grey'])
sp.summary()
```

I tried raw.get_data, but there is an error that says 0 cannot be divided.

**Reply by umair-hassan** [reply]
The you will have to pick the channel of your intetrest using raw data and give it to spindle_detection function in the same way as the text file was getting it. You could perhaps attach one sample data set on drive or here directly and I may give it a try ?

Best, Umair

**Reply by PhD-GOAT** [reply]
GitHub is not allowed to upload the edf file. SC4001E0-PSG. I'm going to link the edf. (https://physionet.org/content/sleep-edfx/1.0.0/sleep-cassette/#files-panel

**Reply by umair-hassan** [reply]
Was the issue solved? Just checking in before I jump on it again. Thanks!

**Reply by PhD-GOAT** [reply]
No, yasa says Spectral Analysis works, but it doesn't seem to be able to visualize or anything else. Although it is possible to read the raw data, edf, it is thought that there is a problem in the algorithm. If want to use it for Yasa, have to handle the edf file, which makes no sense.
